core/string/ustring.cpp

```diff
@@ -2097,11 +2097,7 @@ String::String(const StrRange &p_range) {
 	copy_from(p_range.c_str, p_range.len);
 }
 
-int64_t String::hex_to_int(bool p_with_prefix) const {
-	if (p_with_prefix && length() < 3) {
```

Comment:
Won't removing the `length()` check mean that you might wrongly unreference `s[1]` on a string of length 1?

Also not sure why there weren't checks before accessing `s[0]` for the sign, are we sure that strings are always de length >= 1? Or are we just happily indexing into adjacent memory? That's a pre-existing potential issue but might be worth handling too.

Reply:
Haven't tested yet (still compiling) but I suspect this might now crash if you try `"0".hex_to_int()`.

This should fix it:
```diff
diff --git a/core/string/ustring.cpp b/core/string/ustring.cpp
index 64171cfce3..0b6311dadf 100644
--- a/core/string/ustring.cpp
+++ b/core/string/ustring.cpp
@@ -2098,6 +2098,11 @@ String::String(const StrRange &p_range) {
 }
 
 int64_t String::hex_to_int() const {
+	int len = length();
+	if (len == 0) {
+		return 0;
+	}
+
 	const char32_t *s = ptr();
 
 	int64_t sign = s[0] == '-' ? -1 : 1;
@@ -2106,7 +2111,7 @@ int64_t String::hex_to_int() const {
 		s++;
 	}
 
-	if (s[0] == '0' && s[1] == 'x') {
+	if (len > 2 && s[0] == '0' && s[1] == 'x') {
 		s += 2;
 	}
 
@@ -2134,6 +2139,11 @@ int64_t String::hex_to_int() const {
 }
 
 int64_t String::bin_to_int() const {
+	int len = length();
+	if (len == 0) {
+		return 0;
+	}
+
 	const char32_t *s = ptr();
 
 	int64_t sign = s[0] == '-' ? -1 : 1;
@@ -2142,7 +2152,7 @@ int64_t String::bin_to_int() const {
 		s++;
 	}
 
-	if (s[0] == '0' && s[1] == 'b') {
+	if (len > 2 && s[0] == '0' && s[1] == 'b') {
 		s += 2;
 	}
 
```

Reply:
Confirmed that it crashes without my fix.

Reply:
Good catch. I tried to produce any other crashes with small inputs but I wasn't able to. The string "0x" won't increase `s` at the start, so it reads "0x" as digits, and when it reads the `x` it returns zero - which is fine because this number may as well be zero anyway. The string "-0x" will increase `s` by 3, and then it's false at the start of the `while` so it returns `-1 * 0`.

If you want I could backport a fix to 3.2 for zero-length strings with `p_with_prefix` as false, even though the parameter is not exposed, and true is used in most places.

Reply:
Yeah I think it might be worth fixing in `3.2`, it can probably be triggered in C++ modules or GDNative.